syntax = "proto3";

package cakelab.arflow_grpc.v1;

import "cakelab/arflow_grpc/v1/vector2.proto";
import "cakelab/arflow_grpc/v1/vector2_int.proto";
import "google/protobuf/timestamp.proto";

option csharp_namespace = "CakeLab.ARFlow.Grpc.V1";

message CameraFrame {
  /// https://docs.unity3d.com/Packages/com.unity.xr.arfoundation@6.0/api/UnityEngine.XR.ARSubsystems.XRCpuImage.Format.html
  enum Format {
    FORMAT_UNSPECIFIED = 0;
    // @exclude The number in each field should match the enum XRCpuImage.Format for more convenient conversion.
    // @exclude FORMAT_ANDROID_YUV_420_888 = 1;
    // @exclude FORMAT_IOS_YP_CBCR_420_8BI_PLANAR_FULL_RANGE = 2;
    // @exclude FORMAT_ONECOMPONENT8 = 3;
    // @exclude FORMAT_DEPTHFLOAT32 = 4;
    // @exclude FORMAT_DEPTHUINT16 = 5;
    // @exclude FORMAT_ONECOMPONENT32 = 6;
    // @exclude FORMAT_ARGB32 = 7;
    FORMAT_RGBA32 = 8;
    // @exclude FORMAT_BGRA32 = 9;
    FORMAT_RGB24 = 10;
  }

  /* @exclude
   * // https://docs.unity3d.com/Packages/com.unity.xr.arfoundation@6.0/api/UnityEngine.XR.ARSubsystems.XRCpuImage.Plane.html
   * message Plane {
   *   int32 row_stride = 1;
   *   int32 pixel_stride = 2;
   *   bytes data = 3;
   * }
   */

  /// https://docs.unity3d.com/Packages/com.unity.xr.arfoundation@6.0/api/UnityEngine.XR.ARSubsystems.XRCameraIntrinsics.html
  message Intrinsics {
    Vector2 focal_length = 1;
    Vector2 principal_point = 2;
    Vector2Int resolution = 3;
  }

  google.protobuf.Timestamp device_timestamp = 1;
  double image_timestamp = 2;
  Format format = 3;
  // @exclude repeated Plane planes = 5;
  Intrinsics intrinsics = 4;
  bytes data = 5;
}
